### ISSUE TITLE
NNue interface with (SF minus NNue)

### ISSUE THREAD

**Issue by Dboingue** [Dboingue/Hello-Stockfish#1]
This is only to figure out exactly how NNue position evaluation calls are made in SF without NNue.

Only about the hybridization of this input-output black box (for this exercise) function from position space to some 1D (assumed) scoring function.  The NN part and its optimization art compared to full NN approaches like a0 or lc0, is well documented.  But I could only find one page actually referring to that type of concern. (must be obvious to AB engine developpers and curious forum participants).  Not to me, as I am more of the NN frame of mind by background.  So learning on many fronts. but aiming for least amount of details to understand the algorithms not the code.  However, I came to the conclusion that the code is finally where I should look. hoping not to trip on technicalities of github, or C or C++.

I  am actually hoping to the contrary, that both github and C++ have mechanisms that make it easy to figure out the call graph of a set of modules within SF. so that I could directly get all NNue entry points into the rest of SF modules. (I should make a separate issue asking the universe about that).

The fiber-bundle of questions center:
Given the DFS, exhaustive, AB pruning, min-max algorithm skeleton (+hash table?), and its node classification (tree search dependent) without NNue, what needs to be changed in there to accomodate that NNue function.  **How does NNue module interact with the existing node classes, their existing tests and any static evaluations, including if it shares input with the classical heuristic static position evaluation function.**

The chessprogramming wiki has this statement dating from 2020, about the first SF with hybridization.
[Stockfish_NNUE Hybrid](https://www.chessprogramming.org/Stockfish_NNUE#Hybrid
)
_> (This page was last edited on **20 July 2021,** at 10:44.)_
> **In August 2020** a new patch changed Stockfish NNUE into a hybrid engine: 
> it uses NNUE evaluation only on quite balanced material positions, otherwise uses the classical one. 
> It could speed up to 10% and gain 20 Elo [35]. At that point, NNUE helped to increase already around 100 Elo for Stockfish. 
> In the same month, Stockfish changed the default mode of using evaluation functions from classic to hybrid one, the last step to completely accept NNUE.

Well that is the only place where I could find any thing like that.  I am now wanting to check if that statement is still valid by looking into the code.  Given the lack of public interest or proposal in that NNue-to-SF11 docking scheme (short for non-NNue part of SF), I am not sure that this would not change across SF12, SF13, SF14 and future development.

Debating where I should go systematic. on pull on some weaving thread from a text search with "eval" and "NNue" in vertical proximity.  some grep like utility on github. or locally on my file system form a clone.

The systematic would be a sure way to keep witnessing the evolution of that information.  If such watching workflow could be made, and given that this does not seem like often requested information, that would be a documentation independent way to watch the evolution of the hybridization process of that initially rather exhaustive search engine type.

**Comment by Dboingue** [comment]
Not clear yet:
It seems that the virtues of better early accurate filtering of candidates (partial, only quiescence prediction I bet) is nibbling at the "fuite en avant" to catch that proof of a favourable solution (without caring for any other pruned earlier, even with a non-accurate heuristic).  editorial hypothesis.

Current plan:  
Currently, the lazy assumption is that the above statement still holds.  That there is not direct score evaluation interaction with NNue. The only effect of NNue, is having there is in how it shapes the skeleton search algorithm I mentioned above that which is the same as in SF11 (barring other unrelated optimizations).  By hearding the branching at interior node with a better quiescence prediction  (some fog here as to what kind of quiescence is tested, my ignorance of chess positions possibilities may be at play, in that I assume that there is no art in quiescence testing.  that one can decide without an statistical function, that a postion is quiescent or not).  I hunch that the notion of quiescence itself may have been loosened.  Possiblly, NNue could also be used in pruning.  As I am not clear yet on the exact nature of pruning, it appears to me to be analogous to quiescence testing in that both will weight on branchiing decision, all resulting in which remote node will get a heuristic static eval.

But, it seems that as long as not sharing any input with the heurtistic eval, it is safe to assume that the pedagogical use of SF in engine analyses on SF can be interepreted as SF11 could be.  Until somebody find a way to compute probable accuracy errors from that heuristic (And the error in accuracy from tree search pruning or other branching generation decisions), one can only look at the heuristic static evaluations for the positions at end of PV output for current position, to interpret the score of the root position where the engine analysis was requested.

**Comment by Dboingue** [comment]
`... uses NNue evaluation only on quite balanced material positions, otherwise uses the classical one ...
`

I had the wrong interpretation.  **WARNING**.  Never mind the previous message. (kept for transparency of process).
### Current corrected understanding skeleton

1. Both functions, NNue and the classical SF static position evaluation are using the leaf nodes of the search that is reached upon traditional unchanged scheme of AB pruning, min-max, extensible "horizon". 
3. A new _category_ of leaf nodes has been introduced.  Positions with small material imbalance (count). This might imply 2 things: 
   * a new _test_ function for evaluating such case. i.e. a criterion.  (least cost computation)
   * a parameter to act as _threshold_ ("quite balanced"), 
     * below which NNue provides the static evaluation score, and 
     * above which the classical SF (=SF11*) evaluation only.

For the new position test for material neutrality, from what I heard or read:
- NNue heavy.  
- SF11*  light 
- other?  

hypotheses:
- SF11*, always as first eval, since material count is dominating (was in the past),  and if below threshold for quite balanced, then the costly NNue.  Need knowledge, can't deduce or guess.
- Threshold autotuned or otherwise globally optimized (after NNue training?)
- order of NN training versus whole SF12-14, autotuning (or better now that NN training is better known) of SF11*
- 2 autotuning global optimization mode resulting in 2 sets of parameters values for the positional unknowns of  SF11* class of functions.  borrowing from CS speak instead of mathematics: the 2 parameters sets provide for 2 instances of the heuristic function.  That would solve an issue I had read about, that was concerned with possibility that the classical mode would be neglected and using the same parameter set as that with NNue on.   The 2 instances, seem like the most reasonable.  As there is obvious interaction between the two families of static evaluators (NNue, and SF11*), the autotuning , being done toward maximizing some pipeline of engineXengine tests.  ELO making the cut. 

WIP.  many questions just popped out. and I need to regroup there by topic.  be back soon.

**Comment by Dboingue** [comment]
https://github.com/official-stockfish/Stockfish#a-note-on-classical-evaluation-versus-nnue-evaluation

**Comment by Dboingue** [comment]
Which parameters get tuned differently now that NNue is in the picture?  There was related issue in origin. about neglecting the classical evaluation by not autotuning (a.k.a. afaik, global optimization, too vague, it could include training, which autotuning is not careful about generalization and overfitting (and possibly optimizing toward the test set sampling biases, and getting increasing better at it, but only for the test set, but for a more general chess challenge for example).  Maybe this is not the case anymore. but i have not read explicit confirmation that the auto-tuning in  SF, is doing some cross-validation in the fishnet pipeline.. if that is where all of it happens.

- [ ] gathers fishnet and tuning code here
- [ ] find which parameters in static eval are being from human knowledge, or tuned.  e.g. material count is in design., positional parameters are not, as there is only qualtiative feature human knowledge, (0,1 being the trivial ordering, quant?).
- [ ]  (i.e. drilling down the variable in the source code of SF, until variable is assigned a constant or a valued function output.   once mapping between dev version and exec version is available, centralize all parameters that represent the information of the the classical eval.
- [ ] same for tuning
- [ ] first measure if Sisiphean ambitiion

**Comment by Dboingue** [comment]
compiling SF on windows. aiming at high level AND exact model (math and chess) of SF as input-output function of chess positions to a float score and a successor position to each such input position, in its simplest form.  This could also be vectorize into multi PV versions.  Current strategy is a mixture of black box (FEN in, SF UCI out in console), white box (source code, including comments).  

But given the long evolution of SF, and the many atomic elements of the search and static evaluations, and their intricacies through recursive algorithm implementations, a breadth-first or high level gap has to be first filled.  This is now more apparent as missing in our understanding of what SF does from the user point of view (going in toward exactitude but minimal implementation details, i.e. high level necessary and sufficient when at math/chess level).  

Example of what may be missing in user point of view of my understanding (and yes, this could be my abilities, but i give it my best, and I am not alone in trying to figure it out).

My personal model of SF searches, based of a flexible horizon at 22 plies, with lots on interior nodes between 1 and 22 without evaluation **conflicts** with the command line output experiments info output when throttling down CPU (and to think of it, lichess SF real time engine thinking progress giving x/22 increasing and settling to 22 after a while).  This slow CPU insight finally convinced us to go the extra km (mile) and compile SF, and use debug as a bridge in our strategy.  Also, the lack of surface in the UCI interface in those experiments.

**Comment by Dboingue** [comment]
Now I had to realize that **iterative deepening**, from the human analysis of the engine use as a human chess analysis tool, is not just an optimization for time constraints (although it may have been there for that competition dimension of engine X engine stratospheres tournaments).  From a human point of view, there is no hope of thinking huge sub-trees with only remote static evaluations.  What is the point of a score at current position, that can only be visible at long distance in the search.  None.  Even if that static evaluation was perfectly predicting the winning odds from that position information as starting position, humans need information perceptible within their effective horizon (say).   Humans can count, hence the popularity of material counting system, that is applied by humans (and SF classical static eval. but that is only half the question).

So while my model of desert of static evaluation in the root dependent 22 plies search with rare leaf nodes being actually evaluated, is not so wrong, as it does not changes the final result:  Yes the root attributed score is from heuristic position information at the tip of the variation, which ply depth should be at least 21 (If enough time to compute, no previous entries in the hash table to override it).  21 or 22, depends on ply index versus ply depths conventions X definitions (start with 0 or 1?).

But in order to get there, there have been many more static evaluations.   Not too many too early, though, because of the cost of these compared to more branch explorations allowing more possibilities of material gain further.  This means that the real time iterative deepening contains early information of human interest.  Not just a time shaving optimization.  From what I read, there seems to be a trade off that make this short depths cat-like exploration into further depth, more efficient (on some average) than  the "interior static evaluation desert" model.  That is where i am at.    What is the extent of peppered static evaluations?

What this means also, is that the transposition hash table, is also not just a time shaving optimization, again with respect to human analysis of engine score and PV output.  It is a witness of those static evaluations and the history of branch exploration, dismissal, continuation, or exact score for current iterative depth**.  (I also guess or infer from some of source code reading, that it also keeps track of upper and lowers bounds from past search).  Another missing dimension of the UCI interface, which only gives a signal when full.  From this current tentative understanding (model), update from new information in our strategy of discovery of high level but exact algorithm, I am now concerned with understanding, testing, or verifying a few new things.

**Comment by Dboingue** [comment]
New questions from updated high level model of SF algorithm as implemented today.

types of depths in new model.
- input depth
- iterative deepening current depth
- quiescence search depth, best PV depth to tip (score=tip eval))

SF UCI output only contains 2 output variables/parameter names: 
- depth=input depth
- seldepth= output depth representing ?
- - iterative 
- - quiescence extension
- - quiescence extension for each iterative, at time of info output

Clearly some information of human analysis interest about interpreting SF analysis, is missing.
Argues for debugging again. command line experiments, with throttled CPU speed, can give some idea. 
(memo complete that thought, with some more live testing now that i cleared up my confusion).

Another new thing in my understanding as user looking inward (but carefully, only using source code details to fill in the gaps, or discrepancies between current model of the algorithm, experiments, and available non source code documentation).

- quantitative level of interaction between the hash table*, and iterative deepening.
- quantitative level of interaction between the hash table and AB pruning
- quantitative level of interaction between iterative deepening and AB pruning
- quantitative level of interaction between: All 3 things.
* (assuming only cache of full evals by SF)

Hash table populating dynamics and usage dynamics become something to focus on toward clearing up the fog at high level, and could ties up all the intricacies of the source code making SF time constraint efficient, but making source code as documentation impossible (for me at least, not a novel). A sequence of lines that depict a threaded execution (trees at least), without losing an iota of efficiency to maintainable modularity of the elements we now are considering.

If we can slow motion the Hash table dynamics, we might actually get the full statistical picture and untangling of SF source code into something satisfying.  source code modifications, tip toeing on some TRACE modifications to diagnose hypotheses from our strategy.

**Comment by Dboingue** [comment]
Now i know what i am doing. (not the objective though).  having cleared what my questions are, and having a muse to hunt, will get me through the pain of procedural command line and IDE set up and acrobatic environment configuration battlefield.  

And the debugging will be about the model not my syntax.  If that  TRACE/NO_TRACE C++ feature is that modular and non-invasive mechanism of centralizing all new variable output into one file, with minimal code change on the rest of code.  

I will make a branch here to prevent updating of that.  no intent on pushing upstream. fork for sandbox.  i can keep the master updated in case. but i should protect my new branch.  hopefully that branch will not interfere with anybody else 

(part of the mountain to climb, is also understanding what i just wrote...)

**Comment by Dboingue** [comment]
I am temporary closing this issue. As partly answered from my discussions with a collaborator (source of the progress shown here).
I may finish documenting what let us to the new questions about understanding the evolution of SF parameters and their tuning status across versions.  I have created 2 new issues from here, stemming from more SF command line experiments with UCI interface and more than one computer.  So this issue is only not going to progress until the others do.
